src/Simulation/TargetDefinitions/Decompositions/Measure.qs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.Intrinsic {
+    open Microsoft.Quantum.Diagnostics;
+
+    /// # Summary
+    /// Performs a joint measurement of one or more qubits in the
+    /// specified Pauli bases.
+    ///
+    /// # Description
+    /// The output result is given by the distribution:
+    /// \begin{align}
+    ///     \Pr(\texttt{Zero} | \ket{\psi}) =
+    ///         \frac12 \braket{
+    ///             \psi \mid|
+    ///             \left(
+    ///                 \boldone + P_0 \otimes P_1 \otimes \cdots \otimes P_{N-1}
+    ///             \right) \mid|
+    ///             \psi
+    ///         },
+    /// \end{align}
+    /// where $P_i$ is the $i$th element of `bases`, and where
+    /// $N = \texttt{Length}(\texttt{bases})$.
+    /// That is, measurement returns a `Result` $d$ such that the eigenvalue of the
+    /// observed measurement effect is $(-1)^d$.
+    ///
+    /// # Input
+    /// ## bases
+    /// Array of single-qubit Pauli values indicating the tensor product
+    /// factors on each qubit.
+    /// ## qubits
+    /// Register of qubits to be measured.
+    ///
+    /// # Output
+    /// `Zero` if the $+1$ eigenvalue is observed, and `One` if
+    /// the $-1$ eigenvalue is observed.
+    ///
+    /// # Remarks
+    /// If the basis array and qubit array are different lengths, then the
+    /// operation will fail.
+    @EnableTestingViaName("Test.TargetDefinitions.Measure")
+    operation Measure (bases : Pauli[], qubits : Qubit[]) : Result {
+        CheckQubitUniqueness(qubits);
+        if (Length(bases) != Length(qubits)) { fail "Arrays 'bases' and 'qubits' must be of the same length."; }
+        mutable res = One;
+        if( Length(bases) == 1 ) {
+            within {
+                MapPauli(qubits[0], PauliZ, bases[0]);
+            }
+            apply {
+                set res = M(qubits[0]);
+                PreparePostM(res, qubits[0]);
+            }
+        }
+        else {
+            using( q = Qubit() ) {
```

Comment:
This looks to use an additional spare qubit when measuring a multi-qubit Pauli operator; that's completely valid, but there's a tradeoff between that and finding the right Clifford to map back to a single-qubit measurement, then calling the measurement within that Clifford.

Reply:
Good point! The hope here is that as we find alternative decompositions for certain operations, we can introduce those here and allow users to select between them by choosing different targeting packages. How we can do that in a way that doesn't cause an explosion in the number of targeting packages is another question though...